services/QuillLMS/app/controllers/snapshots_controller.rb

```diff
@@ -77,6 +60,32 @@ def options
     return render json: { error: 'user is not authorized for all specified schools' }, status: 403
   end
 
+  private def retrieve_cache_or_enqueue_worker(worker)
+    return { error: 'unrecognized query type for this endpoint' } unless worker::QUERIES.keys.include?(snapshot_params[:query])
+
+    previous_start, current_start, current_end = Snapshots::Timeframes.calculate_timeframes(snapshot_params[:timeframe],
+      snapshot_params[:timeframe_custom_start],
+      snapshot_params[:timeframe_custom_end])
+    cache_key = cache_key_for_timeframe(previous_start, current_start, current_end)
+    response = Rails.cache.read(cache_key)
+
+    return response if response
+
+    worker.perform_async(cache_key,
+      snapshot_params[:query],
+      current_user.id,
+      {
+        name: snapshot_params[:timeframe],
+        previous_start: previous_start,
+        current_start: current_start,
+        current_end: current_end
+      },
+      snapshot_params[:school_ids],
+      snapshot_params[:grades])
+
+    { message: 'Generating snapshot' }
+  end
+
   private def cache_key_for_timeframe(previous_start, current_start, current_end)
 
     Snapshots::CacheKeys.generate_key(snapshots_params[:query],
```

Comment:
Is `snapshots_params[:query]` a typo?  If so, I would suggest maybe creating a new method:

```
private def query
  snapshot_params[:query]
end
```

since it's used in a few places in this controller.